### DISCUSSION TITLE
Keras 3 and Transformers Not compatable?

### DISCUSSION THREAD

**Discussion by JosephPerst** [community/community#118713] [Programming Help]
### Body

Hey all, I am very confused at what is wrong with my code. I have been trying for hours to figure this out to no avail. I am getting this error when attempting to use Keras 3 and transformers to perform NLP tasks. I have pasted the error below. The only time the code runs is when I use the string 'Adam' within the optimizer section at the bottom. However, I need to be able to adjust my learning rate, and Adam does not seem to work, and i have tried other optimizers. I also attached my code to this post. Keep in mind I am new, and I apologize if my conduct here is out of the ordinary. Thank you in advance!

import pandas as pd
import re
import numpy as np
import tensorflow as tf
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
import nltk
from transformers import LongformerTokenizer, TFLongformerForSequenceClassification
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

# Adjusting Keras imports to use tensorflow.keras
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.utils import to_categorical

# Download necessary NLTK data
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

# Define text cleaning function
def nltk_clean_text(text):
    text = re.sub(r'http\S+|www\S+|https\S+', '', str(text), flags=re.MULTILINE)
    text = re.sub(r'\@\w+|\#', '', str(text))
    tokens = word_tokenize(text)
    tokens = [w.lower() for w in tokens]
    words = [word for word in tokens if word.isalpha()]
    stop_words = set(stopwords.words('english'))
    words = [w for w in words if not w in stop_words]
    lemmatizer = WordNetLemmatizer()
    lemmatized = [lemmatizer.lemmatize(word) for word in words]
    return ' '.join(lemmatized)

# Define rating categorization function
def categorize_rating(rating):
    rating = int(rating)
    if rating > 6:
        return 'positive'
    elif rating == 6:
        return 'neutral'
    else:
        return 'negative'

# Load and preprocess the dataset
df = pd.read_csv(r"C:\Users\jpers\Desktop\NLP\USE THIS FOR NLP 2.csv")
df['rating_category'] = df['rating'].apply(categorize_rating)
df['review'] = df['review'].apply(nltk_clean_text)
df['Extracted Information'] = df['Extracted Information'].apply(nltk_clean_text)
df['condition'] = df['condition'].apply(nltk_clean_text)
df['drug_name_x'] = df['drug_name_x'].apply(nltk_clean_text)
df['combined_text'] = df['review'] + " " + df['drug_name_x'] + " " + df['rating_category']

# Initialize tokenizer and model
tokenizer = LongformerTokenizer.from_pretrained('allenai/longformer-base-4096')
model = TFLongformerForSequenceClassification.from_pretrained('allenai/longformer-base-4096', num_labels=3)

# Encode reviews
def encode_reviews_bigbird(reviews, tokenizer):
    return tokenizer.batch_encode_plus(
        reviews.to_list(),
        padding='max_length',
        truncation=True,
        max_length=1024,
        return_tensors='tf'
    )

# Encode the data
df['rating_category_encoded'] = LabelEncoder().fit_transform(df['rating_category'])
labels = tf.keras.utils.to_categorical(df['rating_category_encoded'], num_classes=3)
X_train, X_test, y_train, y_test = train_test_split(df['combined_text'], labels, test_size=0.2, random_state=42)
X_train_enc = encode_reviews_bigbird(X_train, tokenizer)
X_test_enc = encode_reviews_bigbird(X_test, tokenizer)

# Prepare datasets
train_dataset = tf.data.Dataset.from_tensor_slices(({"input_ids": X_train_enc['input_ids'], "attention_mask": X_train_enc['attention_mask']}, y_train))
test_dataset = tf.data.Dataset.from_tensor_slices(({"input_ids": X_test_enc['input_ids'], "attention_mask": X_test_enc['attention_mask']}, y_test))

# Compile and fit the model
optimizer = Adam(learning_rate=0.001)
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(train_dataset.shuffle(len(X_train)).batch(4), epochs=3, batch_size=4, validation_data=test_dataset.batch(4))

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
File [c:\Users\jpers\AppData\Local\Programs\Python\Python312\Lib\site-packages\transformers\activations_tf.py:22](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/activations_tf.py:22)
     [21](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/activations_tf.py:21) try:
---> [22](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/activations_tf.py:22)     import tf_keras as keras
     [23](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/activations_tf.py:23) except (ModuleNotFoundError, ImportError):

ModuleNotFoundError: No module named 'tf_keras'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
File [c:\Users\jpers\AppData\Local\Programs\Python\Python312\Lib\site-packages\transformers\utils\import_utils.py:1472](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/utils/import_utils.py:1472), in _LazyModule._get_module(self, module_name)
   [1471](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/utils/import_utils.py:1471) try:
-> [1472](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/utils/import_utils.py:1472)     return importlib.import_module("." + module_name, self.__name__)
   [1473](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/utils/import_utils.py:1473) except Exception as e:

File [c:\Users\jpers\AppData\Local\Programs\Python\Python312\Lib\importlib\__init__.py:90](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/importlib/__init__.py:90), in import_module(name, package)
     [89](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/importlib/__init__.py:89)         level += 1
---> [90](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/importlib/__init__.py:90) return _bootstrap._gcd_import(name[level:], package, level)

File <frozen importlib._bootstrap>:1387, in _gcd_import(name, package, level)

File <frozen importlib._bootstrap>:1360, in _find_and_load(name, import_)
...
   [1476](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/utils/import_utils.py:1476)         f" traceback):\n{e}"
   [1477](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/utils/import_utils.py:1477)     ) from e

RuntimeError: Failed to import transformers.models.longformer.modeling_tf_longformer because of the following error (look up to see its traceback):
Your currently installed version of Keras is Keras 3, but this is not yet supported in Transformers. Please install the backwards-compatible tf-keras package with `pip install tf-keras`

File <frozen importlib._bootstrap>:1360, in _find_and_load(name, import_)
...
   [1476](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/utils/import_utils.py:1476)         f" traceback):\n{e}"
   [1477](file:///C:/Users/jpers/AppData/Local/Programs/Python/Python312/Lib/site-packages/transformers/utils/import_utils.py:1477)     ) from e

RuntimeError: Failed to import transformers.models.longformer.modeling_tf_longformer because of the following error (look up to see its traceback):
Your currently installed version of Keras is Keras 3, but this is not yet supported in Transformers. Please install the backwards-compatible tf-keras package with `pip install tf-keras`.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by JadenYangCA** [comment]
same error when using tensorflow-cpu==2.17.0 and python 3.12

**Reply by charleybin** [reply]
same error

**Reply by JadenYangCA** [reply]
I solved this error using the following combo:
pip install tensorflow-cpu==2.16.1
pip install tf-keras==2.16.0 --no-dependencies

**Comment by baslia** [comment]
Hello, is there a way to track when transformers would be compatible with keras v3 ?

**Comment by manfredkremer** [comment]
I am also very puzzled by this problem, especially as in google colab the transformers do not complain about keras version 3.8. But on my computer this combination invariably crashes.
